src/core-sdk-tasks/DownloadFile.cs

```diff
@@ -39,31 +54,70 @@ public override bool Execute()
                 var filePath = Uri.Substring(FileUriProtocol.Length);
                 Log.LogMessage($"Copying '{filePath}' to '{DestinationPath}'");
                 File.Copy(filePath, DestinationPath);
+                return true;
             }
-            else
+
+            var lastStatus = await DownloadWithRetries(Uri, DestinationPath);
+            if (lastStatus == false)
             {
-                Log.LogMessage(MessageImportance.High, $"Downloading '{Uri}' to '{DestinationPath}'");
+                if (!string.IsNullOrEmpty(PrivateUri))
+                {
+                    lastStatus = await DownloadWithRetries(PrivateUri, DestinationPath);
+                }
+            }
+
+            return lastStatus == true;
+        }
 
+        /// <summary>
+        /// Attempt to download file from `source` with retries when response error is different of FileNotFound and Success.
+        /// </summary>
+        /// <param name="source">URL to the file to be downloaded.</param>
+        /// <param name="target">Local path where to put the downloaded file.</param>
+        /// <returns>true: Download Succeeded. false: Download failed with 404. null: Download failed but is retriable.</returns>
+        private async Task<bool?> DownloadWithRetries(string source, string target)
+        {
+            Random rng = new Random();
+
+            Log.LogMessage(MessageImportance.High, $"Attempting download '{source}' to '{target}'");
+
+            for (int retryNumber = 0; retryNumber < MaxRetries; retryNumber++)
+            {
                 using (var httpClient = new HttpClient())
                 {
-                    var getTask = httpClient.GetStreamAsync(Uri);
-
                     try
                     {
-                        using (var outStream = File.Create(DestinationPath))
+                        var httpResponse = await httpClient.GetAsync(source);
+
+                        // The Azure Storage REST API returns '400 - Bad Request' in some cases
+                        // where the resource is not found on the storage.
+                        // https://docs.microsoft.com/en-us/rest/api/storageservices/common-rest-api-error-codes
+                        if (httpResponse.StatusCode == HttpStatusCode.NotFound ||
+                            httpResponse.ReasonPhrase.IndexOf("The requested URI does not represent any resource on the server.", StringComparison.OrdinalIgnoreCase) >= 0)
                         {
-                            getTask.Result.CopyTo(outStream);
+                            Log.LogMessage($"File not found error while attempting download using the following URL(s): '{source}'.");
+                            return false;
```

Comment:
What happened to the plan to add diagnosability here?

Reply:
Done: https://dnceng.visualstudio.com/internal/_build/results?buildId=411924&view=results